{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d4175cf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load all the corresponding packages\n",
    "from Bio import Entrez\n",
    "from Bio import Medline\n",
    "import pandas as pd\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed\n",
    "from itertools import islice\n",
    "import multiprocessing\n",
    "from urllib.error import HTTPError\n",
    "from collections import Counter\n",
    "from gensim.parsing.preprocessing import preprocess_string,strip_tags, strip_punctuation, remove_stopwords\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bd7ee6be",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Functions to retrieve Pubmed Data \n",
    "def doc_function(url, count = 0):\n",
    "    \"\"\"\n",
    "    doc_function that retrieves pubmed articles from the queried term\n",
    "    input: search query\n",
    "    output: dictionary of:\n",
    "    abstracts, titles, journals, PMIDS\n",
    "    \"\"\"\n",
    "\n",
    "    # this can be abstracted to a dictionary\n",
    "    # try block to retrieve the data for the query\n",
    "    try:\n",
    "        Entrez.email = \"maximilian.zeidler@i-med.ac.at\"\n",
    "        esearch_query = Entrez.esearch(db=\"pubmed\", term=url,retmax = 200000)\n",
    "        esearch_result = Entrez.read(esearch_query)\n",
    "        count = esearch_result['Count']\n",
    "        esearch_query1 = Entrez.esearch(db=\"pubmed\", term=url, retmax = count, retmode = \"xlm\")\n",
    "        esearch_result1 = Entrez.read(esearch_query1)\n",
    "        idlist = esearch_result1[\"IdList\"]\n",
    "        handle = Entrez.efetch(db=\"pubmed\", id=idlist, rettype=\"medline\", retmode=\"text\", retmax = count)\n",
    "        records = Medline.parse(handle)\n",
    "        chunked_records = chunk(records, 20)\n",
    "        records_dataframe_list = progress_records_multiprocessing(chunked_records)\n",
    "        final_dataframe = append_dataframes(records_dataframe_list)\n",
    "        return final_dataframe\n",
    "    except HTTPError:\n",
    "        count += 1\n",
    "        time.sleep(10)\n",
    "        if count < 5:\n",
    "            doc_function(url, count)\n",
    "        else:\n",
    "            raise ConnectionError\n",
    "            print(\"No connection can be established\")\n",
    "            \n",
    "def retrieve_records(records):\n",
    "    \"\"\"\n",
    "    This function retrieves the abstracts, id and title from the \n",
    "    Records.\n",
    "    The input is a generator which will be evualted lazy per record\n",
    "    \n",
    "    \"\"\"\n",
    "    print(\"new Thread\")\n",
    "    corpus_dictionary = {\"ID\": [], \n",
    "                         \"abstract\": [],\n",
    "                         \"Title\":[], \n",
    "                         \"Journal\": [],\n",
    "                         \"Publication_date\":[], \n",
    "                         \"first\":[], \n",
    "                         \"last\": []\n",
    "                        }\n",
    "    for record in records:\n",
    "        try:\n",
    "            if ((\"DP\") in record.keys() and (\"PMID\") in record.keys() and (\"AB\") in record.keys() and (\"TI\") in record.keys() and (\"JT\") in record.keys() and (\"FAU\") in record.keys()):\n",
    "                corpus_dictionary[\"ID\"].append(record[\"PMID\"])\n",
    "                corpus_dictionary[\"abstract\"].append(record[\"AB\"])\n",
    "                corpus_dictionary[\"Title\"].append(record[\"TI\"])\n",
    "                corpus_dictionary[\"Journal\"].append(record[\"JT\"])\n",
    "                corpus_dictionary[\"Publication_date\"].append(record[\"DP\"])\n",
    "                corpus_dictionary[\"first\"].append(record[\"FAU\"][0])\n",
    "                corpus_dictionary[\"last\"].append(record[\"FAU\"][-1])\n",
    "            else:\n",
    "                print(\"Searched record is not available\")\n",
    "                pass\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    record_dataframe = pd.DataFrame(corpus_dictionary)\n",
    "    return record_dataframe\n",
    "    \n",
    "def progress_records_multiprocessing(records_list, workers = 8):\n",
    "        \"\"\"Runs the Worker Thread for each chunk and returns an iterator of the list\n",
    "        results of the records fetched from the PubMed Outline\n",
    "        Args:\n",
    "            new_list (_type_): _description_\n",
    "        Returns:\n",
    "            _type_: _description_\n",
    "        \"\"\"\n",
    "        with ThreadPoolExecutor(max_workers=workers) as executor:\n",
    "            return executor.map(retrieve_records, records_list, timeout=60)\n",
    "\n",
    "def chunk(it, size):\n",
    "        \"\"\"List will be chunked into equally sized parts\n",
    "        Args:\n",
    "            it (list): will be the iterable of the list\n",
    "            size (int): describes the number of chunks based on size per chunk\n",
    "        Returns:\n",
    "            iter : iterable object for all chunks (Generator)\n",
    "        \"\"\"\n",
    "        it = iter(it)\n",
    "        return iter(lambda: tuple(islice(it, size)), ())\n",
    "    \n",
    "def append_dataframes(records:list) -> pd.DataFrame:\n",
    "    \"\"\" Goes through the dataframes and appends all the chunked\n",
    "    dataframes to one final dataframe\n",
    "    \n",
    "    \"\"\"\n",
    "    final_records = pd.DataFrame()\n",
    "    for i in records:\n",
    "        final_records = pd.concat([final_records, i], axis = 0)\n",
    "    return final_records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "da6f0d33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "Searched record is not available\n",
      "Searched record is not available\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n",
      "new Thread\n"
     ]
    }
   ],
   "source": [
    "# retrieve the search term\n",
    "url = \"scRNA sequencing Brain\"\n",
    "records_dataframes = doc_function(url) # get the abstracts from pubmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "be0cc334",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>abstract</th>\n",
       "      <th>Title</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Publication_date</th>\n",
       "      <th>first</th>\n",
       "      <th>last</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36621073</td>\n",
       "      <td>Brain metastases (BMs) of lung cancer are comm...</td>\n",
       "      <td>The clinical application of (68)Ga-PSMA PET/CT...</td>\n",
       "      <td>Translational oncology</td>\n",
       "      <td>2023 Jan 6</td>\n",
       "      <td>Pei, Yuchen</td>\n",
       "      <td>Gao, Yang</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>36619671</td>\n",
       "      <td>Severe traumatic spinal cord injury (SCI) lead...</td>\n",
       "      <td>Neurogenesis potential of oligodendrocyte prec...</td>\n",
       "      <td>Frontiers in cellular neuroscience</td>\n",
       "      <td>2022</td>\n",
       "      <td>Zhao, Qing</td>\n",
       "      <td>Xie, Ning</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>36612109</td>\n",
       "      <td>The pituitary gland is one of the most cellula...</td>\n",
       "      <td>Transcriptomic Profiles of Normal Pituitary Ce...</td>\n",
       "      <td>Cancers</td>\n",
       "      <td>2022 Dec 24</td>\n",
       "      <td>Oh, Jun Y</td>\n",
       "      <td>Aghi, Manish K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>36594818</td>\n",
       "      <td>Emerging evidence suggests that the meningeal ...</td>\n",
       "      <td>The meningeal transcriptional response to trau...</td>\n",
       "      <td>eLife</td>\n",
       "      <td>2023 Jan 3</td>\n",
       "      <td>Bolte, Ashley C</td>\n",
       "      <td>Lukens, John R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>36583014</td>\n",
       "      <td>As single-cell chromatin accessibility profili...</td>\n",
       "      <td>Benchmarking automated cell type annotation to...</td>\n",
       "      <td>Frontiers in genetics</td>\n",
       "      <td>2022</td>\n",
       "      <td>Wang, Yuge</td>\n",
       "      <td>Zhao, Hongyu</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID                                           abstract  \\\n",
       "0  36621073  Brain metastases (BMs) of lung cancer are comm...   \n",
       "1  36619671  Severe traumatic spinal cord injury (SCI) lead...   \n",
       "2  36612109  The pituitary gland is one of the most cellula...   \n",
       "3  36594818  Emerging evidence suggests that the meningeal ...   \n",
       "4  36583014  As single-cell chromatin accessibility profili...   \n",
       "\n",
       "                                               Title  \\\n",
       "0  The clinical application of (68)Ga-PSMA PET/CT...   \n",
       "1  Neurogenesis potential of oligodendrocyte prec...   \n",
       "2  Transcriptomic Profiles of Normal Pituitary Ce...   \n",
       "3  The meningeal transcriptional response to trau...   \n",
       "4  Benchmarking automated cell type annotation to...   \n",
       "\n",
       "                              Journal Publication_date            first  \\\n",
       "0              Translational oncology       2023 Jan 6      Pei, Yuchen   \n",
       "1  Frontiers in cellular neuroscience             2022       Zhao, Qing   \n",
       "2                             Cancers      2022 Dec 24        Oh, Jun Y   \n",
       "3                               eLife       2023 Jan 3  Bolte, Ashley C   \n",
       "4               Frontiers in genetics             2022       Wang, Yuge   \n",
       "\n",
       "             last  \n",
       "0       Gao, Yang  \n",
       "1       Xie, Ning  \n",
       "2  Aghi, Manish K  \n",
       "3  Lukens, John R  \n",
       "4    Zhao, Hongyu  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get overview\n",
    "records_dataframes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8e224d1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 482 entries, 0 to 3\n",
      "Data columns (total 7 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   ID                482 non-null    object\n",
      " 1   abstract          482 non-null    object\n",
      " 2   Title             482 non-null    object\n",
      " 3   Journal           482 non-null    object\n",
      " 4   Publication_date  482 non-null    object\n",
      " 5   first             482 non-null    object\n",
      " 6   last              482 non-null    object\n",
      "dtypes: object(7)\n",
      "memory usage: 30.1+ KB\n",
      "None\n",
      "              ID                                           abstract  \\\n",
      "count        482                                                482   \n",
      "unique       482                                                482   \n",
      "top     36621073  Brain metastases (BMs) of lung cancer are comm...   \n",
      "freq           1                                                  1   \n",
      "\n",
      "                                                    Title  \\\n",
      "count                                                 482   \n",
      "unique                                                482   \n",
      "top     The clinical application of (68)Ga-PSMA PET/CT...   \n",
      "freq                                                    1   \n",
      "\n",
      "                                    Journal Publication_date         first  \\\n",
      "count                                   482              482           482   \n",
      "unique                                  224              376           460   \n",
      "top     The Journal of biological chemistry             2022  Kobayashi, S   \n",
      "freq                                     24               20             7   \n",
      "\n",
      "            last  \n",
      "count        482  \n",
      "unique       422  \n",
      "top     Anzai, K  \n",
      "freq           9  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Counter({'Translational oncology': 1,\n",
       "         'Frontiers in cellular neuroscience': 3,\n",
       "         'Cancers': 1,\n",
       "         'eLife': 14,\n",
       "         'Frontiers in genetics': 8,\n",
       "         'bioRxiv : the preprint server for biology': 1,\n",
       "         'Molecular neurodegeneration': 2,\n",
       "         'Neuroscience': 2,\n",
       "         'Genome research': 4,\n",
       "         'Molecular immunology': 1,\n",
       "         'Cell reports methods': 1,\n",
       "         'Methods in molecular biology (Clifton, N.J.)': 2,\n",
       "         'Journal of neurochemistry': 4,\n",
       "         'Annals of clinical and translational neurology': 1,\n",
       "         'Frontiers in cell and developmental biology': 3,\n",
       "         'Pathology, research and practice': 1,\n",
       "         'Nucleic acids research': 11,\n",
       "         'Pharmaceuticals (Basel, Switzerland)': 1,\n",
       "         \"Alzheimer's research & therapy\": 1,\n",
       "         'Journal of the American Society of Nephrology : JASN': 1,\n",
       "         'Frontiers in immunology': 8,\n",
       "         'Journal of cancer research and therapeutics': 1,\n",
       "         'The Journal of neuroscience : the official journal of the Society for Neuroscience': 1,\n",
       "         'Diabetes, metabolic syndrome and obesity : targets and therapy': 1,\n",
       "         'Oncogene': 3,\n",
       "         'Aging and disease': 3,\n",
       "         'Aging cell': 1,\n",
       "         'BMC biology': 2,\n",
       "         'Frontiers in neuroscience': 4,\n",
       "         'Briefings in bioinformatics': 4,\n",
       "         'Journal of neuroinflammation': 5,\n",
       "         'Computational and structural biotechnology journal': 1,\n",
       "         'International journal of obesity (2005)': 1,\n",
       "         'Seminars in cell & developmental biology': 1,\n",
       "         'Cell reports': 9,\n",
       "         'CNS neuroscience & therapeutics': 1,\n",
       "         'Biochimica et biophysica acta. Molecular basis of disease': 1,\n",
       "         'EBioMedicine': 1,\n",
       "         'STAR protocols': 4,\n",
       "         'Stem cell reports': 4,\n",
       "         'Frontiers in molecular neuroscience': 2,\n",
       "         'Oxidative medicine and cellular longevity': 2,\n",
       "         'Scientific data': 2,\n",
       "         'Brain sciences': 1,\n",
       "         'Cells': 3,\n",
       "         'Archives of toxicology': 1,\n",
       "         'Autophagy': 1,\n",
       "         'Journal of computational biology : a journal of computational molecular cell biology': 1,\n",
       "         'Frontiers in endocrinology': 1,\n",
       "         'Scientific reports': 8,\n",
       "         'PLoS genetics': 2,\n",
       "         'FASEB journal : official publication of the Federation of American Societies for Experimental Biology': 2,\n",
       "         'Nature communications': 19,\n",
       "         'Redox biology': 1,\n",
       "         'Journal of leukocyte biology': 1,\n",
       "         'Neural regeneration research': 1,\n",
       "         'Advanced science (Weinheim, Baden-Wurttemberg, Germany)': 1,\n",
       "         'Cell & bioscience': 1,\n",
       "         'Molecular psychiatry': 4,\n",
       "         'GigaScience': 1,\n",
       "         'Neuroscience research': 1,\n",
       "         'Cell systems': 4,\n",
       "         'Brain and behavior': 1,\n",
       "         'Neuro-oncology': 2,\n",
       "         'Proceedings of the National Academy of Sciences of the United States of America': 18,\n",
       "         'Frontiers in nutrition': 1,\n",
       "         'Cellular and molecular life sciences : CMLS': 2,\n",
       "         'Disease models & mechanisms': 1,\n",
       "         'Genome biology': 7,\n",
       "         'BMB reports': 1,\n",
       "         'BMC genomics': 5,\n",
       "         'Database : the journal of biological databases and curation': 1,\n",
       "         'Nature': 2,\n",
       "         'Oncoimmunology': 1,\n",
       "         'Clinical and translational medicine': 3,\n",
       "         'Molecular metabolism': 1,\n",
       "         'The Journal of clinical investigation': 1,\n",
       "         'Cancer letters': 1,\n",
       "         'Cell': 6,\n",
       "         'BMC bioinformatics': 3,\n",
       "         'Journal of visualized experiments : JoVE': 1,\n",
       "         'Genomics, proteomics & bioinformatics': 2,\n",
       "         'American journal of human genetics': 1,\n",
       "         'Protein & cell': 3,\n",
       "         'European neuropsychopharmacology : the journal of the European College of Neuropsychopharmacology': 1,\n",
       "         'Molecular neurobiology': 3,\n",
       "         'Frontiers in aging neuroscience': 1,\n",
       "         'The EMBO journal': 2,\n",
       "         'Cell stem cell': 1,\n",
       "         'Blood': 1,\n",
       "         'Frontiers in oncology': 2,\n",
       "         'Journal of cerebral blood flow and metabolism : official journal of the International Society of Cerebral Blood Flow and Metabolism': 2,\n",
       "         'Annual review of vision science': 1,\n",
       "         'JCI insight': 1,\n",
       "         'RNA (New York, N.Y.)': 2,\n",
       "         'Development (Cambridge, England)': 1,\n",
       "         'Journal of genetics and genomics = Yi chuan xue bao': 1,\n",
       "         'Frontiers in neuroanatomy': 1,\n",
       "         'Cancer cell': 1,\n",
       "         'Nature biotechnology': 2,\n",
       "         'Communications biology': 1,\n",
       "         'PLoS computational biology': 6,\n",
       "         'Science advances': 1,\n",
       "         'PLoS medicine': 1,\n",
       "         'Clinical cancer research : an official journal of the American Association for Cancer Research': 2,\n",
       "         'Developmental cell': 3,\n",
       "         'Bioinformatics (Oxford, England)': 7,\n",
       "         'Cell reports. Medicine': 1,\n",
       "         'Science (New York, N.Y.)': 2,\n",
       "         'iScience': 2,\n",
       "         'Nature metabolism': 1,\n",
       "         'Journal of ethnopharmacology': 1,\n",
       "         'International journal of molecular sciences': 1,\n",
       "         'Cerebral cortex (New York, N.Y. : 1991)': 1,\n",
       "         'eNeuro': 3,\n",
       "         'Mathematical biosciences and engineering : MBE': 1,\n",
       "         'Wiley interdisciplinary reviews. RNA': 1,\n",
       "         'Med (New York, N.Y.)': 1,\n",
       "         'Frontiers in cardiovascular medicine': 1,\n",
       "         'The FEBS journal': 1,\n",
       "         'Current protocols': 1,\n",
       "         'RNA biology': 3,\n",
       "         'International journal of environmental research and public health': 1,\n",
       "         'International journal of infectious diseases : IJID : official publication of the International Society for Infectious Diseases': 1,\n",
       "         'Experimental dermatology': 1,\n",
       "         'Molecular cell': 2,\n",
       "         'Viruses': 1,\n",
       "         'Biology of sex differences': 1,\n",
       "         'Brain pathology (Zurich, Switzerland)': 1,\n",
       "         'Aging': 2,\n",
       "         'Advances in experimental medicine and biology': 1,\n",
       "         'Journal of cellular physiology': 2,\n",
       "         'Epigenomes': 1,\n",
       "         'Cell death & disease': 2,\n",
       "         'Neuron': 5,\n",
       "         'Gut': 1,\n",
       "         'medRxiv : the preprint server for health sciences': 1,\n",
       "         'BMC medical genetics': 1,\n",
       "         'Neurobiology of disease': 2,\n",
       "         'Journal of Cancer': 1,\n",
       "         'Frontiers in bioengineering and biotechnology': 1,\n",
       "         'Biomolecules': 1,\n",
       "         'Current pharmaceutical design': 1,\n",
       "         'Neuropharmacology': 1,\n",
       "         'Nature genetics': 1,\n",
       "         'Nature methods': 2,\n",
       "         'Acta neuropathologica communications': 1,\n",
       "         'Genes': 2,\n",
       "         'Quantitative biology (Beijing, China)': 1,\n",
       "         'Trends in neurosciences': 1,\n",
       "         'Current opinion in biotechnology': 1,\n",
       "         'Nature medicine': 1,\n",
       "         'PloS one': 5,\n",
       "         'BMC systems biology': 1,\n",
       "         \"Child's nervous system : ChNS : official journal of the International Society for Pediatric Neurosurgery\": 1,\n",
       "         'Nature protocols': 2,\n",
       "         'PLoS biology': 1,\n",
       "         'Molecular brain': 1,\n",
       "         'ACS chemical neuroscience': 1,\n",
       "         'The Journal of biological chemistry': 24,\n",
       "         'Current biology : CB': 1,\n",
       "         'Brain : a journal of neurology': 2,\n",
       "         'Journal of neurovirology': 1,\n",
       "         'Trends in molecular medicine': 1,\n",
       "         'Epilepsia': 1,\n",
       "         'Journal of dental research': 1,\n",
       "         'F1000Research': 1,\n",
       "         'Current opinion in neurobiology': 1,\n",
       "         'Neuropathology and applied neurobiology': 1,\n",
       "         'Cell death and differentiation': 1,\n",
       "         'International journal of molecular medicine': 1,\n",
       "         'Circulation research': 1,\n",
       "         'ACS nano': 1,\n",
       "         'BMC infectious diseases': 2,\n",
       "         'Neuropathology : official journal of the Japanese Society of Neuropathology': 1,\n",
       "         'Brain & development': 1,\n",
       "         'Gene': 4,\n",
       "         \"Journal of Alzheimer's disease : JAD\": 1,\n",
       "         'Acta neuropathologica': 1,\n",
       "         'The Journal of pathology': 1,\n",
       "         'Hematology (Amsterdam, Netherlands)': 1,\n",
       "         'Journal of toxicology and environmental health. Part A': 1,\n",
       "         'Molecular and cellular biology': 7,\n",
       "         'Brain research': 2,\n",
       "         'Molecular and cellular neurosciences': 1,\n",
       "         'Endocrinology': 2,\n",
       "         'Cellular signalling': 1,\n",
       "         'Neoplasia (New York, N.Y.)': 1,\n",
       "         'Glia': 1,\n",
       "         'The Journal of pharmacology and experimental therapeutics': 2,\n",
       "         'Cell metabolism': 1,\n",
       "         'Molecular endocrinology (Baltimore, Md.)': 2,\n",
       "         'Methods in enzymology': 1,\n",
       "         'Trends in biochemical sciences': 1,\n",
       "         'Journal of molecular biology': 2,\n",
       "         'Gene expression patterns : GEP': 1,\n",
       "         'Virology': 1,\n",
       "         'Gene therapy': 1,\n",
       "         'Drug metabolism and pharmacokinetics': 1,\n",
       "         'Human molecular genetics': 2,\n",
       "         'Neuroreport': 1,\n",
       "         'Biochemical and biophysical research communications': 8,\n",
       "         'Histochemistry and cell biology': 1,\n",
       "         'Genomics': 2,\n",
       "         'Biochimica et biophysica acta': 4,\n",
       "         'The Journal of cell biology': 4,\n",
       "         'Journal of virology': 1,\n",
       "         'American journal of respiratory cell and molecular biology': 1,\n",
       "         'Biotechnic & histochemistry : official publication of the Biological Stain Commission': 1,\n",
       "         'DNA and cell biology': 2,\n",
       "         'Mechanisms of development': 1,\n",
       "         'Laboratory investigation; a journal of technical methods and pathology': 1,\n",
       "         'Immunology and cell biology': 1,\n",
       "         'The Biochemical journal': 1,\n",
       "         'Annals of the New York Academy of Sciences': 1,\n",
       "         'Molecular biology reports': 1,\n",
       "         'Molecular pharmacology': 1,\n",
       "         'Cell and tissue research': 1,\n",
       "         'European journal of biochemistry': 1,\n",
       "         'Cell growth & differentiation : the molecular biology journal of the American Association for Cancer Research': 1,\n",
       "         'Journal of cell science': 1,\n",
       "         'Journal of immunology (Baltimore, Md. : 1950)': 1,\n",
       "         'Journal of autoimmunity': 1,\n",
       "         'Developmental biology': 2})"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this is important to get an overview over missing data and the inferred data types per column\n",
    "print(records_dataframes.info())\n",
    "# gets overview over the most frequent data \n",
    "# also usually used to retrieve min, max, mean for each column if numeric, int, float\n",
    "print(records_dataframes.describe())\n",
    "# exploratory data analysis\n",
    "# using counter object from the collections package\n",
    "top_journals = Counter(records_dataframes[\"Journal\"].tolist())\n",
    "top_journals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e9d1b7a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "The Journal of biological chemistry                                                24\n",
       "Nature communications                                                              19\n",
       "Proceedings of the National Academy of Sciences of the United States of America    18\n",
       "eLife                                                                              14\n",
       "Nucleic acids research                                                             11\n",
       "                                                                                   ..\n",
       "Science advances                                                                    1\n",
       "PLoS medicine                                                                       1\n",
       "Cell reports. Medicine                                                              1\n",
       "Nature metabolism                                                                   1\n",
       "International journal of molecular sciences                                         1\n",
       "Name: Journal, Length: 224, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# getting directly from the pandas package\n",
    "top_journals_df = records_dataframes[\"Journal\"].value_counts()\n",
    "top_journals_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7c13c61a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# task evaluate it using your own function !\n",
    "journals_list = records_dataframes[\"Journal\"].tolist()\n",
    "def counter(record_list, dataframe = False):\n",
    "    \"\"\"\n",
    "    Function to count the occurence of each word\n",
    "    args:\n",
    "        record_list type(list) <- list that holds the words to count\n",
    "        dataframe (bool) <- true if converted into pd.Series\n",
    "    returns\n",
    "        dict or series\n",
    "    \"\"\"\n",
    "    counter_journals = {}\n",
    "    for i in record_list:\n",
    "        # check the base state\n",
    "        if i not in counter_journals: \n",
    "            # can also used the keys methods here\n",
    "            counter_journals.update({i:1})\n",
    "        else:\n",
    "            counter_journals[i] += 1\n",
    "            \n",
    "    if dataframe:\n",
    "        counter_dataframe = pd.Series(counter_journals).sort_values(ascending = False)\n",
    "        return counter_dataframe\n",
    "    return counter_journals\n",
    "\n",
    "journal_dict = counter(journals_list)\n",
    "journal_dataframe = counter(journals_list, dataframe = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1a4d07a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question: How can we calculate from this the percentage?\n",
    "frequency_dataframe = pd.DataFrame(journal_dataframe, columns = [\"count\"])\n",
    "# Answer:\n",
    "frequency_dataframe[\"frequency\"] = (frequency_dataframe[\"count\"]/ frequency_dataframe[\"count\"].sum()) * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4eb0d70b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "of            3891\n",
       "and           3793\n",
       "in            2610\n",
       "to            1748\n",
       "a             1596\n",
       "that          1170\n",
       "with           853\n",
       "is             852\n",
       "for            773\n",
       "cell           758\n",
       "by             608\n",
       "we             601\n",
       "from           556\n",
       "RNA            532\n",
       "cells          525\n",
       "expression     491\n",
       "are            467\n",
       "The            464\n",
       "was            430\n",
       "dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Repurpose counter function for word frequency?\n",
    "# thats why writing function/class methods is important\n",
    "# We reduce workload, error rates and increase efficency\n",
    "# Let's look at the Table for the abstracts\n",
    "def records_tolist(records: list, preprocessed = False, stopwords: list = False) -> pd.Series:\n",
    "    \"\"\" \n",
    "    Of importance here is that you understand what you are doing \n",
    "    Functions are available gettting rid of stopwords \n",
    "    Removing Punctuations\n",
    "    Introducing Lemmtization and more \n",
    "    Which is all important for NLP\n",
    "    \n",
    "    args:\n",
    "        records type(list of lists) <- holds the word records\n",
    "        stopwords(list) <- default None, holds the \n",
    "    \"\"\"\n",
    "    if preprocessed:\n",
    "        record_flatten = records\n",
    "        \n",
    "    else:\n",
    "        record_flatten = [i.replace(\"(\", \"\").replace(\")\",\"\").split(\" \") for i in records.tolist()]\n",
    "        \n",
    "    if stopwords:\n",
    "        # check if is in stopword\n",
    "        record_flatten = [t.lower() for i in record_flatten for t in i]\n",
    "        record_flatten = [i for i in record_flatten if i not in stopwords]\n",
    "        \n",
    "    else:    \n",
    "        #not removing stopwords\n",
    "        record_flatten = [t for i in record_flatten for t in i]\n",
    "        \n",
    "    word_series = counter(record_flatten, dataframe = True)\n",
    "    return word_series\n",
    "\n",
    "word_series = records_tolist(records_dataframes[\"abstract\"])\n",
    "word_series.iloc[1:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ed18175f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocessing_list(liste,stopword_list):\n",
    "    \"\"\" \n",
    "    preprocess the abstract list, remove stopwords, punctuations, numbers\n",
    "    input: \n",
    "    liste: list of abstracts\n",
    "    stopword_list: list of stopwords\n",
    "    returns list of list <- holding the abstracts\n",
    "    \"\"\"\n",
    "    processed_abstracts = []\n",
    "    for i in liste:\n",
    "        CUSTOM_FILTERS = [lambda x: x.lower(),remove_stopwords]\n",
    "        a = preprocess_string(i, CUSTOM_FILTERS)\n",
    "        no_integers = [x for x in a if not (x.isdigit() or x[0] == '-' and x[1:].isdigit())]\n",
    "        no_integers = [x for x in no_integers if x not in stopword_list]\n",
    "        no_integers = [re.sub(\"[^A-Za-z0-9|-]\",\"\",x) for x in no_integers]\n",
    "        processed_abstracts.append(no_integers)\n",
    "    return processed_abstracts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "01f3ea5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cell           771\n",
       "rna            532\n",
       "scrna-seq      449\n",
       "single-cell    414\n",
       "brain          411\n",
       "this           392\n",
       "these          350\n",
       "data           334\n",
       "human          298\n",
       "sequencing     287\n",
       "which          275\n",
       "have           257\n",
       "our            243\n",
       "analysis       235\n",
       "using          204\n",
       "types          185\n",
       "their          184\n",
       "mouse          184\n",
       "has            183\n",
       "dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Self created list of stopwords\n",
    "# Usually not the way to go, but to gain understanding of how to process the data\n",
    "stopwords = [\"the\",\"of\",\"in\",\"to\",\n",
    "             \"with\",\"were\",\"a\",\"was\",\n",
    "             \"by\",\"the\", \"the\",\"that\",\n",
    "             \"for\",\"is\",\"as\",\"on\",\"significant\",\n",
    "             \"among\",\"although\",\"especially\",\"kg\",\n",
    "             \"km\",\"mainly\",\"ml\",\"mm\",\n",
    "             \"disease\",\"significantly\",\"obtained\",\"mutation\",\n",
    "             \"significant\",\"quite\",\"result\",\"results\",\"estimated\",\n",
    "             \"interesting\",\"conducted\",\"associated\",\"performed\",\n",
    "             \"respectively\",\"larger\",\"genes\",\"gene\", \"mutations\",\n",
    "             \"related\",\"expression\",\"pattern\",\"mutation\",\"clc\",\"identified\",\n",
    "             \"suprisingly\",\"preferentially\",\"subsequently\",\"far\",\"little\",\n",
    "             \"known\",\"importantly\",\"synonymous\",\"skipping\",\"father\",\n",
    "             \"mother\",\"pedigree\",\"novo\",\"rescues\",\"rescued\",\"restored\",\n",
    "             \"exhibits\",\"induce\", \"Background\",\"Objective\",\"Methods\",\n",
    "             \"cells\", \"kinase\",\"activation\",\"protein\",\n",
    "             \"be\",\"at\", \"we\", \"p\",\"from\",\"or\",\"after\",\"treatment\",\n",
    "             \"=\", \"are\",\".\", \"an\"\n",
    "            ]\n",
    "word_series_no_stop = records_tolist(records_dataframes[\"abstract\"], stopwords = stopwords)\n",
    "word_series_no_stop.iloc[1:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "20514631",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cell           796\n",
       "rna            582\n",
       "brain          548\n",
       "scrna-seq      498\n",
       "data           449\n",
       "single-cell    414\n",
       "sequencing     309\n",
       "analysis       307\n",
       "human          302\n",
       "types          244\n",
       "neurons        210\n",
       "mouse          196\n",
       "study          181\n",
       "methods        172\n",
       "cellular       170\n",
       "small          166\n",
       "development    159\n",
       "molecular      157\n",
       "expressed      155\n",
       "neuronal       151\n",
       "dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Check with preprocessing\n",
    "records_dataframes[\"processed_abstract\"] = preprocessing_list(records_dataframes[\"abstract\"].tolist(),stopwords)\n",
    "word_series_prep = records_tolist(records_dataframes[\"processed_abstract\"], True, stopwords = stopwords)\n",
    "word_series_prep[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32f59c2a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "vscode": {
   "interpreter": {
    "hash": "b29e83ee11e7f59cb38c31a3bdf4f8395742e175511a4c2dd46f06d53a300255"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
